packages/next/src/server/app-render/app-render.tsx

```diff
@@ -336,90 +336,157 @@ function createFlightDataResolver(ctx: AppRenderContext) {
   }
 }
 
-type ServerComponentsRendererOptions = {
+type ReactServerAppProps = {
+  tree: LoaderTree
   ctx: AppRenderContext
   preinitScripts: () => void
-  options: ServerComponentRendererOptions
+  asNotFound: boolean
 }
+async function ReactServerApp({
+  tree,
+  ctx,
+  preinitScripts,
+  asNotFound,
+}: ReactServerAppProps) {
+  preinitScripts()
+  // Create full component tree from root to leaf.
+  const injectedCSS = new Set<string>()
+  const injectedJS = new Set<string>()
+  const injectedFontPreloadTags = new Set<string>()
+  const {
+    getDynamicParamFromSegment,
+    query,
+    providedSearchParams,
+    appUsingSizeAdjustment,
+    componentMod: { AppRouter, GlobalError },
+    staticGenerationStore: { urlPathname },
+  } = ctx
+  const initialTree = createFlightRouterStateFromLoaderTree(
+    tree,
```

Comment:
I've extracted the Render and Error RSC components into static functions accepting props rather than constructed functions consuming a closure. The actual contents on what is rendered should not have changed (apologies for how annoying this will be to check in review)